eeps/eep-0058.md

```diff
@@ -0,0 +1,255 @@
+  boolean result: when it returns `false` the current element is
+  skipped from being passed to "comprehension"; when it results
+  in `true`, the element is passed to "comprehension".
+
+So, in this document we are going to keep this separation, but
+sometimes calling the whole combination of comprehension, filter and
+generator as just "comprehension", because generator syntax is not
+used outside of the comprehension context.
+
+
+### Syntax
+
+Proposed syntax for map comprehension part is
+
+    '#{' KeyExpr '=>' ValueExpr '||' GeneratorsFilters '}'
+
+Where `GeneratorsFilters` are one or more list/map/binary generator
+and/or filter, separated by comma.
+
+Proposed syntax for map generator part is
+
+    KeyPattern ':=' ValuePattern '<-' MapOrIteratorExpr
+
+Example combination of map comprehension and map generator
+
+    #{Key => Value || Key := Value <- MapOrIterator}
+
+Map comprehension may be combined with list or binary generator and
+vice-versa
+
+    [{Key, Value} || Key := Value <- MapOrIterator]
+    #{K => K || K <- List}
+    #{K => V || <<K, V>> <= Binary}
+
+`MapOrIterator` is any Erlang expression that evaluates to `map()`
+or `maps:iterator()` datatype.
+
+### Semantics
+
+Map comprehension should be semantically equivalent to the
+`maps:from_list/1` call over the list comprehension producing
+the list of 2-tuples:
+
+    #{K => V || ...}
+
+equivalent to
+
+    maps:from_list([{K, V} || ...])
+
+
+Map generator should be semantically equivalent to the list generator
+running over the result of `maps:to_list/1`, when the input is a
+map datatype:
+
+    #{ ... || K := V <- Map}
+
+equivalent to
+
+    #{ ... || {K, V} <- maps:to_list(Map)}
+
+And for `maps:iterator()` as input - it can't be easily expressed as
+equivalent list generator. But the idea is that it's equivalent to
+the consumption of the iterator by repetitive calls to `maps:next/1`
+and matching `Key` and `Value` over key and value patterns
+accordingly on each iteration loop.
+
+Map generator should raise `error({bad_generator, ExprResult})`
+if the `MapOrIterator` expression evaluation produced anything but
+map or map iterator.
+
+Key pattern is allowed to be catch-all / contain unbound variables
+(which is not allowed in other map pattern-matching contexts).
+
+Variable binding and shadowing rules should be the same as for list
+and binary comprehensions (which follows from the equivalence rules
+described above).
+
+Rationale
+=========
+
+### Syntax for arrow, generator pattern and comprehension key-value pair
+
+We decided to go with the syntax proposed in EEP-43.
+
+`#{ .. }` were chosen for the outer tokens of map comprehension
+because the same are used when constructing a new map - rule similar to
+list and binary comprehensions.
+
+`K := V <-` is not ambiguous for the parser, so unlike binary
+generator, there is no need for a new arrow. `<-` were chosen over
+`<=` because latter is less visually distinguishable when it is
```

Comment:
```suggestion
generator, there is no need for a new arrow. `<-` was chosen over
`<=` because the latter is less visually distinguishable when it is
```